Simple Rearrangement

RLee64 opened the pull request:
## Overview
Just enough C++ code that gets the ``x = y + z`` example working (where x and z are defined).

## Changes
``CMakeLists.txt`` - For some reason symengine and gmp start complaining and issue a ton of warnings that get treated as errors. I couldn't manage to fix the problem, so I've temporarily commented out a couple lines that seem to control that functionality (feels wrong to do it, but I already spent more than an hour trying to fix it). A quick google search suggested that these errors might be due to differences between 64-bit and 32-bit, but I ran ``dumpbin /headers`` on the dlls and they all came back as 64-bit, so I feel inclined to believe the root cause is something else.
Snippet of the errors:
```
C:\gmp\include\gmp.h(1792): error C2220: the following warning is treated as an error
C:\gmp\include\gmp.h(1792): warning C4244: 'return': conversion from 'mp_limb_t' to 'unsigned long', possible loss of data
C:\gmp\include\gmp.h(2237): warning C4146: unary minus operator applied to unsigned type, result still unsigned
```

``src/CMakeLists.txt`` - Added hardcoded symengine and gmp dependencies. Unfortunately, Python bindings couldn't be created from this alone, so I've done manual testing through C++ for the time being. Also, the symengine and gmp dlls weren't able to be found by the libcellml testing library (even though installing and using the library in another project works perfectly fine?). I've tried adding their respective bins to my env path and also moved the installation out from Program Files directly to C: but no luck there. Right now I've managed to get things working by pasting the dlls into the build/tests folder.

``rearrange()`` - A method for `AnalyserInternalEquation`, if successful, it returns a rearranged ast, otherwise it returns a nullptr. This should be the main point of access to the new symengine rearrangement functionality from existing code.

``getSymEngineRepresentation()`` - A method for `AnalyserInternalEquation` returns the SymEngine equivalent of its AST. Calls itself recursively while traversing through the AST.

``parseSymEngineExpression()`` - A method for `AnalyserInternalEquation` that converts a SymEngine expression into an AST equivalent.

``Header file changes`` - Had some trouble with getting the respective header files to recognise symengine (#include didn't work). Managed to get it to function with namespace, hopefully this is the correct way of going about it?

``Test file changes`` - Added the x = y + z example to the test suite. The associated test is currently just appended to the end of tests/analyser/analyser.cpp. If I'm planning to have a bunch of tests to cover rearrangement and DAE->ODE conversion, do you think it'd be worth setting up a new test file? It'd also speed up testing times since I could just test that file (as opposed to all ``utils_unit_test_analyser`` test cases)

## Other Notes
* Using symengine seems to make things very verbose, especially when I've got parameters like ``std::map<SymEngine::RCP<const SymEngine::Symbol>, AnalyserInternalVariablePtr, SymEngine::RCPBasicKeyLess> &astMap``, is there good way to shorten this stuff?
* Some existing tests are currently failing so I'll spend some time fixing the issues (on top of fixing anything from your review) before moving on.

RLee64 committed:
early trial (should revert later)

RLee64 committed:
it """builds"""

RLee64 committed:
clean-up and bug fix

RLee64 committed:
Added simple rearrangement test

RLee64 committed:
Fixed naming error on cellml test file

RLee64 committed:
Replaced custom struct with existing RCP comparator

RLee64 committed:
moved symengine conversion to analyserinternalequation

agarny commented on src/api/libcellml/analyserequationast.h:
None of the changes in this file are needed _per se_. If you need `<map>`, you should include it wherever you actually need it. The same when it comes to the `RCP`, `Basic`, and `Symbol` classes. None of them are needed in this header file, so they have nothing to do here.

RLee64 replied:
Ah, thanks for catching that, that code was leftover from when I originally had some implementation here 😅 

Will make sure to remove

agarny commented on src/analyserequationast.cpp:
The two `#include`s are not needed _per se_, so they ought to be removed. You should only `#include` files because you have added some "proper" code to the file and they need things from the `#include` files.

agarny commented on src/CMakeLists.txt:
It's fine to do it this way. That's what we discussed: do whatever it takes to get you going.

Still, this means that, at some point (whether before or after you leave), we will need to update our CMake script to automatically detect SymEngine. As for GMP, I am a bit surprised since I would have expected it to be statically linked to SymEngine. Anyway, something for later to worry about.

agarny commented on tests/analyser/analyser.cpp:
```diff
@@ -1060,3 +1060,16 @@ TEST(Analyser, unsuitablyConstrainedNlaSystem)
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto analyser = libcellml::Analyser::create();
+    analyser->analyseModel(model);
```
It's mainly me, but I like to have an empty line between a declaration and some consuming code.

RLee64 replied:
Gotcha, I'll keep that in mind (honestly, my personal preference so far has been the complete opposite, I usually prefer to group them together so that they look like a nice 'block'). More than happy to change my habits for this project though.

agarny commented on tests/analyser/analyser.cpp:
```diff
@@ -1060,3 +1060,16 @@ TEST(Analyser, unsuitablyConstrainedNlaSystem)
+    analyser->analyseModel(model);
+
+    EXPECT_EQ(libcellml::AnalyserModel::Type::ALGEBRAIC, analyser->analyserModel()->type());
+}
```
That red sign is because GitHub expects files to end with an empty line. This is a convention that we also use. So, yes, you need to add an empty line at the end of the file.

<img width="166" height="141" alt="Image" src="https://github.com/user-attachments/assets/3bce9d3c-b32b-4de9-96f3-9deab1bdfc21" />

RLee64 replied:
Gotcha, will fix

agarny commented on tests/resources/analyser/simple_rearrangement.cellml:
This file is not being used, so it should be deleted.

RLee64 replied:
Ah whoops, looks like this also slipped past me 😅

RLee64 committed:
try catch for parsing ast to symengine

just a blanket block to prevent unrecognisable AST components form passing through

agarny commented on CMakeLists.txt:
If you don't want warnings to be treated as errors then you need to build libCellML using `-DTWAE=OFF`. So, uncomment this and pass the aforementioned definition to CMake when configuring libCellML.

Otherwise, we will clearly need to address those warnings at some point, but it's not the aim of your project, so feel free to use `-DTWAE=OFF` for now.

This being said, and if you have a bit of time and are curious, one way to go around this is by surrounding SymEngine's header with a couple of header files that disable some warnings. For instance (from [here](https://github.com/opencor/libopencor/blob/main/src/misc/compiler.cpp#L30-L38), [here](https://github.com/opencor/libopencor/blob/main/src/3rdparty/LLVMClang/llvmbegin.h), and [here](https://github.com/opencor/libopencor/blob/main/src/3rdparty/LLVMClang/llvmend.h); `BUILDING_USING_MSVC` is specific to libOpenCOR, so in your case you might just want to use `_MSC_VER`):
```c++
#include "llvmbegin.h" // <-- Surrounding header.
#include "llvm/IR/LegacyPassManager.h"
#include "llvm/IR/Module.h"
#include "llvm/MC/TargetRegistry.h"
#include "llvm/Support/Host.h"
#include "llvm/Support/TargetSelect.h"
#include "llvm/Target/TargetMachine.h"
#include "llvm-c/Core.h"
#include "llvmend.h" // <-- Surrounding header.
```
with `llvmbegin.h`:
```c++
...
#if defined(BUILDING_USING_MSVC)
#    pragma warning(push)
#    pragma warning(disable: 4100)
#    pragma warning(disable: 4127)
#    pragma warning(disable: 4146)
#    pragma warning(disable: 4244)
#    pragma warning(disable: 4245)
#    pragma warning(disable: 4267)
#    pragma warning(disable: 4291)
#    pragma warning(disable: 4324)
#    pragma warning(disable: 4389)
#    pragma warning(disable: 4458)
#    pragma warning(disable: 4459)
#    pragma warning(disable: 4624)
#    pragma warning(disable: 4996)
#    pragma warning(disable: 5054)
#elif...
```
and `llvmend.h`:
```c++
...
#if defined(BUILDING_USING_MSVC)
#    pragma warning(pop)
#elif...
```

RLee64 replied:
Thanks for this, -DTWAE=OFF works as expected :)

I've uncommented the lines here - aaf7ebb3f54310fadbdd689d38136fbe2dacbf26

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,129 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
            || variableOnRhs(variable);
 }
 
+SymEngine::RCP<const SymEngine::Basic> AnalyserInternalEquation::getSymEngineRepresentation(AnalyserEquationAstPtr ast, const std::map<std::string, SymEngine::RCP<const SymEngine::Symbol>> &symbolMap)
```
Rename `getSymEngineRepresentation()` to `symEngineRepresentation()`.

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,129 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+    AnalyserEquationAstPtr leftAst = ast->leftChild();
+    AnalyserEquationAstPtr rightAst = ast->rightChild();
+
+    // Recursively call getConvertedAst on left and right children
```
All of our comments end (or, rather, should end) with a full stop.

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,129 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+        // Seems like the voi doesn't exist in mAllVariables, so we don't have an easy means of access
+        // For now we'll just throw an error if the symbol is not found
+        if (symbolMap.find(ast->variable()->name()) == symbolMap.end()) {
+            throw std::runtime_error("Unsupported variable in getSymEngineRepresentation");
```
libCellML doesn't throw exceptions. So, you need to find another solution. Anyway, at some point, you will also need to account for the VOI (for which a test should be created).

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,129 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+    }
+}
+
+AnalyserEquationAstPtr AnalyserInternalEquation::parseSymEngineExpression(SymEngine::RCP<const SymEngine::Basic> &expr,
```
I would rename `expr` with `seExpression` (for SymEngine expression).

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,129 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+        if (children.size() > 1) {
+            ast->setRightChild(parseSymEngineExpression(children[1], astMap));
+        }
+    }
```
We agree that this is wrong. So, a test for more than two children should be created at some point.

agarny commented on src/analyser.cpp:
```diff
@@ -278,6 +402,14 @@ bool AnalyserInternalEquation::check(const AnalyserModelPtr &analyserModel, bool
+        if (newAst != nullptr) {
+            mAst = newAst;
+        }
+    }
```
Remember that you will also need to update the type of some variables and of the equation.

RLee64 replied:
Yeah that makes sense, a bit difficult for me to imagine at this moment what exactly might need to be changed? Hopefully they should make themselves obvious as I start adding more tests.

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,129 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+    return ast;
+}
+
+AnalyserEquationAstPtr AnalyserInternalEquation::rearrange(const AnalyserInternalVariablePtr &variable)
```
Rename `rearrange()` to `rearrangeFor()`.

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,129 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+        SymEngine::RCP<const SymEngine::Symbol> sym = SymEngine::symbol(var->mVariable->name());
+        symbolMap[var->mVariable->name()] = sym;
+        astMap[sym] = var;
+    }
```
Use full variable names, i.e. `variable` rather than `var`, `symbol` rather than `sym`.

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,129 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+        astMap[sym] = var;
+    }
+
+    SymEngine::RCP<const SymEngine::Basic> equation;
```
Rename `equation` to `seEquation`.

agarny replied:
Still relevant.

RLee64 replied:
5d27a2493335756d391bd94c7247a17cc8cc89fc

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,129 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+    }
+
+    SymEngine::RCP<const SymEngine::Set> solutionSet = solve(equation, symbolMap[variable->mVariable->name()]);
+
```
Delete empty line.

agarny commented on src/analyser_p.h:
```diff
@@ -22,6 +24,12 @@ limitations under the License.
 #include "logger_p.h"
 #include "utilities.h"
 
+namespace SymEngine {
+template<class T> class RCP;
+class Basic;
+class Symbol;
+} // namespace SymEngine
+
```
Yes, this time it's right to have those declarations in this header file (and `#include <map>`).

agarny commented on src/analyser_p.h:
```diff
@@ -128,6 +136,10 @@ struct AnalyserInternalEquation
+    AnalyserEquationAstPtr rearrange(const AnalyserInternalVariablePtr &variable);
+    AnalyserEquationAstPtr parseSymEngineExpression(SymEngine::RCP<const SymEngine::Basic> &expr, std::map<SymEngine::RCP<const SymEngine::Symbol>, AnalyserInternalVariablePtr, SymEngine::RCPBasicKeyLess> &astMap);
+    SymEngine::RCP<const SymEngine::Basic> getSymEngineRepresentation(AnalyserEquationAstPtr ast, const std::map<std::string, SymEngine::RCP<const SymEngine::Symbol>> &symbolMap);
+
```
Reverse order to match the order in which they are defined in the `.cpp` file, i.e.
```c++
SymEngine::RCP<const SymEngine::Basic> getSymEngineRepresentation(AnalyserEquationAstPtr ast, const std::map<std::string, SymEngine::RCP<const SymEngine::Symbol>> &symbolMap);
AnalyserEquationAstPtr parseSymEngineExpression(SymEngine::RCP<const SymEngine::Basic> &expr, std::map<SymEngine::RCP<const SymEngine::Symbol>, AnalyserInternalVariablePtr, SymEngine::RCPBasicKeyLess> &astMap);
AnalyserEquationAstPtr rearrange(const AnalyserInternalVariablePtr &variable);
```

agarny reviewed (changes requested):
Well done! That looks very promising!

agarny commented:
> `Header file changes` - Had some trouble with getting the respective header files to recognise symengine (#include didn't work). Managed to get it to function with namespace, hopefully this is the correct way of going about it?

If you are talking about what you did in `src/api/libcellml/analyserequationast.h` then this can be fine but it shouldn't be in this header file.

> `Test file changes` - Added the x = y + z example to the test suite. The associated test is currently just appended to the end of tests/analyser/analyser.cpp. If I'm planning to have a bunch of tests to cover rearrangement and DAE->ODE conversion, do you think it'd be worth setting up a new test file? It'd also speed up testing times since I could just test that file (as opposed to all `utils_unit_test_analyser` test cases)

Yes, I would most probably expect you to create a new file. Maybe a file called `analysersymengine.cpp` to make it clear that this file tests various things that require SymEngine.

> ## Other Notes
> * Using symengine seems to make things very verbose, especially when I've got parameters like `std::map<SymEngine::RCP<const SymEngine::Symbol>, AnalyserInternalVariablePtr, SymEngine::RCPBasicKeyLess> &astMap`, is there good way to shorten this stuff?

Yes, you could define something like this once and reuse it afterwards. For instance:
```c++
using SymEngineAstMap = std::map<SymEngine::RCP<const SymEngine::Symbol>, AnalyserInternalVariablePtr, SymEngine::RCPBasicKeyLess>;
```
and then have:
```c++
SymEngineAstMap &astMap;
```

> * Some existing tests are currently failing so I'll spend some time fixing the issues (on top of fixing anything from your review) before moving on.

100% agreed!

RLee64 committed:
cleaned-up code after review

RLee64 committed:
I forgot a fullstop :(

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,127 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
            || variableOnRhs(variable);
 }
 
+SymEngine::RCP<const SymEngine::Basic> AnalyserInternalEquation::symEngineRepresentation(AnalyserEquationAstPtr ast, const std::map<std::string, SymEngine::RCP<const SymEngine::Symbol>> &symbolMap)
```
Might in fact want to rename `symEngineRepresentation()` to `symEngineEquation().

RLee64 replied:
5ffd2209fa588f7fe4925173c48d0587b4344f83

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,127 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+        // Seems like the voi doesn't exist in mAllVariables, so we don't have an easy means of access.
+        // For now we'll just throw an error if the symbol is not found.
+        if (symbolMap.find(ast->variable()->name()) == symbolMap.end()) {
+            throw std::runtime_error("Unsupported variable in symEngineRepresentation");
```
No throwing of exceptions. Same on line 230.

RLee64 replied:
The exceptions currently let the caller know that a symengine representation couldn't be found, is there a good way we can achieve the same outcome without using them?

agarny replied:
Yes, you could return `SymEngine::null`. So, you could have:
```c++
SymEngine::RCP<const SymEngine::Basic> AnalyserInternalEquation::symEngineEquation(const AnalyserEquationAstPtr &ast, const SymEngineSymbolMap &symbolMap)
{
    if (ast == nullptr) {
        return SymEngine::null;
    }

    AnalyserEquationAstPtr leftAst = ast->leftChild();
    AnalyserEquationAstPtr rightAst = ast->rightChild();

    // Recursively call getConvertedAst on left and right children.
    SymEngine::RCP<const SymEngine::Basic> left = symEngineEquation(leftAst, symbolMap);
    SymEngine::RCP<const SymEngine::Basic> right = symEngineEquation(rightAst, symbolMap);

    if ((left == SymEngine::null) || (right == SymEngine::null)) {
        return SymEngine::null;
    }

    // Analyse mAst current type and value.
    switch (ast->type()) {
    case AnalyserEquationAst::Type::EQUALITY:
        return Eq(left, right);
    case AnalyserEquationAst::Type::PLUS:
        return add(left, right);
    case AnalyserEquationAst::Type::CI:
        // Seems like the voi doesn't exist in mAllVariables, so we don't have an easy means of access.
        if (symbolMap.find(ast->variable()->name()) == symbolMap.end()) {
            return SymEngine::null;
        }
        return symbolMap.at(ast->variable()->name());
    default:
        return SymEngine::null;
    }
}
```
And in `AnalyserInternalEquation::rearrangeFor()`, you could replace:
```c++
    SymEngine::RCP<const SymEngine::Basic> seEquation;
    try {
        seEquation = symEngineEquation(mAst, symbolMap);
    } catch (const std::runtime_error &e) {
        // Our parser was unable to convert the AST to a SymEngine expression.
        return nullptr;
    }
```
with:
```c++
    SymEngine::RCP<const SymEngine::Basic> seEquation = symEngineEquation(mAst, symbolMap);

    if (seEquation == SymEngine::null) {
        // Our parser was unable to convert the AST to a SymEngine expression.
        return nullptr;
    }
```

RLee64 replied:
But isn't it sometimes acceptable to have nullptr children? Leaf nodes (e.g. CI/CN) have no children on either side and even some operators (e.g. +) might only have one child (e.g. +a from the analyseNode examples)
<img width="200" height="185" alt="image" src="https://github.com/user-attachments/assets/ff0ee53d-8f66-4c32-a95a-a10565afde69" />

agarny replied:
Well spotted! Ok, then another solution is to use a `std::tuple`, i.e. have:
```c++
std::tuple<bool, SymEngine::RCP<const SymEngine::Basic>> AnalyserInternalEquation::symEngineEquation(const AnalyserEquationAstPtr &ast, const SymEngineSymbolMap &symbolMap)
{
    if (ast == nullptr) {
        return {true, SymEngine::null};
    }

    AnalyserEquationAstPtr leftAst = ast->leftChild();
    AnalyserEquationAstPtr rightAst = ast->rightChild();

    // Recursively call getConvertedAst on left and right children.
    auto [leftSuccess, left] = symEngineEquation(leftAst, symbolMap);
    auto [rightSuccess, right] = symEngineEquation(rightAst, symbolMap);

    if (!leftSuccess || !rightSuccess) {
        return {false, SymEngine::null};
    }

    // Analyse mAst current type and value.
    switch (ast->type()) {
    case AnalyserEquationAst::Type::EQUALITY:
        return {true, Eq(left, right)};
    case AnalyserEquationAst::Type::PLUS:
        return {true, add(left, right)};
    case AnalyserEquationAst::Type::CI:
        // Seems like the voi doesn't exist in mAllVariables, so we don't have an easy means of access.
        if (symbolMap.find(ast->variable()->name()) == symbolMap.end()) {
            return {false, SymEngine::null};
        }
        return {true, symbolMap.at(ast->variable()->name())};
    default:
        return {false, SymEngine::null};
    }
}
```
and
```c++
    auto [success, seEquation] = symEngineEquation(mAst, symbolMap);

    if (!success) {
        // Our parser was unable to convert the AST to a SymEngine expression.
        return nullptr;
    }
```
Hopefully, `std::tuple` is supported in the standard of C++ that we are using. If not then we might have to bump the version of the standard we are using.

agarny commented on src/analyser.cpp:
```diff
@@ -278,6 +400,14 @@ bool AnalyserInternalEquation::check(const AnalyserModelPtr &analyserModel, bool
                                    mVariables.front() :
                                    nullptr;
 
+    // If we have one variable left, but it's not isolated, try to rearrange it.
+    if ((unknownVariableLeft != nullptr) && !variableOnLhsOrRhs(unknownVariableLeft)) {
+        auto newAst = rearrangeFor(unknownVariableLeft);
+        if (newAst != nullptr) {
+            mAst = newAst;
+        }
+    }
```
Put a note in the comment about having to update the type of some variables and of the equation. You might want to prepend such a note with `TODO`.

RLee64 replied:
c78fdb840a23a5ba48198a80a1c0eead9126301b

agarny commented on src/analyser.cpp:
```diff
@@ -198,6 +199,127 @@ bool AnalyserInternalEquation::variableOnLhsOrRhs(const AnalyserInternalVariable
+
+    // Assume two children max.
+    // This is likely wrong since SYMENGINE_ADD could have x + y + z (and thus 3 children),
+    // but it's sufficient for this very early implementation.
```
Replace this with some kind of a `TODO` comment.

RLee64 replied:
c78fdb840a23a5ba48198a80a1c0eead9126301b

agarny reviewed (commented):
Still some minor changes.

RLee64 committed:
Change symEngineRepresentation to symEngineEquation

RLee64 committed:
Added TODO comments

RLee64 committed:
Renamed equation to seEquation

RLee64 committed:
Renabled cmaketxt warning as error lines

RLee64 committed:
Defined type aliases for symengine maps

Also changed the name of astMap to variableMap to better reflect what it's actually mapping to

RLee64 committed:
Updated function parameters to use const and pass by reference

RLee64 commented:
Also for review two extra commits that don't directly address any of the existing comments
afa335825476ca33470269baf88332e2ed60c8d6 - Type aliasing for maps
a0e333b1e48c9add60c11ce7224690bf09a5b396 - Updated function parameters (so that they're all const and pass by ref)

agarny commented:
> Updated function parameters (so that they're all const and pass by ref)

Thanks for that, I can't believe that I missed it!

RLee64 committed:
Added support for parsing SymEngine integers to AST

RLee64 committed:
Assign parenthood to recreated ast tree

RLee64 committed:
Removed DAE test case and associated files

RLee64 committed:
Replaced exception use in symEngineEquation()

Substituted with std::pair<bool, SymEngine::RCP<const SymEngine::Basic>> use instead

RLee64 committed:
Removed redundant map include

RLee64 committed:
Change SymEngineEquationResult to use tuple instead of pair

agarny commented:
Closing in favour of PR #2.
